core/lib/node/src/resource/resource_collection.rs

```diff
@@ -0,0 +1,106 @@
+///
+/// The purpose of this container is to allow different tasks to register their resource in a single place for some
+/// other task to consume. For example, tasks may register their healthchecks, and then healthcheck task will observe
+/// all the provided healthchecks.
+pub struct ResourceCollection<T: 'static> {
+    /// Collection of the resources.
+    resources: Arc<RwLock<Vec<T>>>,
+    /// Whether someone took the value from this collection.
+    resolved: Arc<AtomicBool>,
+    /// Whether the wiring process has been completed.
+    wired: watch::Receiver<bool>,
+}
+
+impl<T> Clone for ResourceCollection<T> {
+    fn clone(&self) -> Self {
+        Self {
+            resources: self.resources.clone(),
+            resolved: self.resolved.clone(),
+            wired: self.wired.clone(),
+        }
+    }
+}
+
+impl<T> fmt::Debug for ResourceCollection<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("ResourceCollection")
+            .field("resources", &"{..}")
+            .field("resolved", &self.resolved)
+            .field("wired", &self.wired)
+            .finish_non_exhaustive()
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum ResourceCollectionError {
+    #[error("Adding resources to the collection is not allowed after the wiring is complete")]
+    AlreadyWired,
+    #[error("Resource collection is already resolved by the component {0}")]
+    AlreadyResolved(&'static str),
+}
+
+impl<T: 'static> ResourceCollection<T> {
+    pub(crate) fn new(wired: watch::Receiver<bool>) -> Self {
+        Self {
+            resources: Default::default(),
+            resolved: Default::default(),
+            wired,
+        }
+    }
+
+    pub fn push(&self, resource: T) -> Result<(), ResourceCollectionError> {
+        // This check is sufficient, since no task is guaranteed to be running when the value changes.
+        if *self.wired.borrow() {
+            return Err(ResourceCollectionError::AlreadyWired);
+        }
+
+        let mut handle = self.resources.write().unwrap();
+        handle.push(resource);
+        Ok(())
+    }
+
+    pub async fn resolve(mut self) -> Result<Vec<T>, ResourceCollectionError> {
+        // Guaranteed not to hang on server shutdown, since the node will change the value before any task is
+        // actually spawned (per framework rules). For most cases, this check will resolve immediately, unless
+        // some tasks would spawn something from the `IntoZkSyncTask` impl.
+        self.wired
+            .changed()
+            .await
+            .map_err(|_| ResourceCollectionError::AlreadyResolved(std::any::type_name::<T>()))?;
+
+        let mut handle = self.resources.write().unwrap();
+        if self.resolved.load(std::sync::atomic::Ordering::Relaxed) {
+            return Err(ResourceCollectionError::AlreadyResolved(
+                std::any::type_name::<T>(),
+            ));
+        }
+        let resources = std::mem::take(&mut *handle);
```

Comment:
This required a slight refactoring of the healthcheck server, which I initially wanted to avoid, but generally, I agree that it'd be more consistent/predictable. Changed.